File: upload_orders_to_retailcrm.py
#!/usr/bin/env python3
import argparse
import json
import sys
import time
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional, Tuple
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Upload orders from a JSON file to RetailCRM API v5."
    )
    parser.add_argument(
        "--env-file",
        default=".env",
        help="Path to the environment file. Default: .env",
    )
    parser.add_argument(
        "--file",
        default=None,
        help="Path to the orders JSON file. Overrides ORDERS_FILE from env.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Validate and print normalized orders without sending them to RetailCRM.",
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=None,
        help="Only process the first N orders.",
    )
    parser.add_argument(
        "--delay",
        type=float,
        default=None,
        help="Delay between requests in seconds. Overrides REQUEST_DELAY_SECONDS from env.",
    )
    return parser.parse_args()


def load_env(env_path: Path) -> Dict[str, str]:
    if not env_path.exists():
        raise FileNotFoundError(f"Env file not found: {env_path}")

    result: Dict[str, str] = {}
    for raw_line in env_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue

        key, value = line.split("=", 1)
        result[key.strip()] = value.strip().strip('"').strip("'")

    return result


def load_json(path: Path) -> Any:
    for encoding in ("utf-8-sig", "utf-8"):
        try:
            return json.loads(path.read_text(encoding=encoding))
        except UnicodeDecodeError:
            continue

    raise UnicodeDecodeError("utf-8", b"", 0, 1, f"Unable to decode file: {path}")


class RetailCrmApiError(RuntimeError):
    def __init__(self, message: str, details: Dict[str, Any]) -> None:
        super().__init__(message)
        self.details = details


class RetailCrmClient:
    def __init__(self, base_url: str, api_key: str, timeout: int = 30) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def get(self, path: str) -> Dict[str, Any]:
        request = Request(
            f"{self.base_url}{path}",
            headers={
                "X-API-KEY": self.api_key,
                "Accept": "application/json",
            },
        )
        return self._read_json(request)

    def post_form(self, path: str, payload: Dict[str, str]) -> Dict[str, Any]:
        request = Request(
            f"{self.base_url}{path}",
            data=urlencode(payload).encode("utf-8"),
            headers={
                "X-API-KEY": self.api_key,
                "Accept": "application/json",
                "Content-Type": "application/x-www-form-urlencoded",
            },
            method="POST",
        )
        return self._read_json(request)

    def _read_json(self, request: Request) -> Dict[str, Any]:
        try:
            with urlopen(request, timeout=self.timeout) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace")
            try:
                details = json.loads(body)
            except json.JSONDecodeError:
                details = {"errorMsg": body or str(exc)}
            raise RetailCrmApiError(
                f"RetailCRM API returned HTTP {exc.code}: {details}",
                details,
            ) from exc
        except URLError as exc:
            raise RuntimeError(f"Failed to reach RetailCRM API: {exc}") from exc


def pick_site_code(
    explicit_site: Optional[str],
    sites: Dict[str, Dict[str, Any]],
) -> str:
    if explicit_site:
        if explicit_site in sites:
            return explicit_site
        raise ValueError(
            f"Configured RETAILCRM_SITE_CODE '{explicit_site}' is not available."
        )

    if len(sites) == 1:
        return next(iter(sites))

    raise ValueError(
        "RetailCRM account has multiple sites. Set RETAILCRM_SITE_CODE in .env."
    )


def active_codes(reference: Dict[str, Dict[str, Any]]) -> set[str]:
    return {
        code
        for code, item in reference.items()
        if item.get("active", True)
    }


def pick_default_code(reference: Dict[str, Dict[str, Any]]) -> Optional[str]:
    for code, item in reference.items():
        if item.get("defaultForApi"):
            return code

    for code, item in reference.items():
        if item.get("defaultForCrm"):
            return code

    for code, item in reference.items():
        if item.get("active", True):
            return code

    return None


def ensure_number(value: Any, field_name: str) -> float:
    try:
        return float(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"'{field_name}' must be a number, got {value!r}") from exc


def ensure_positive_quantity(value: Any) -> float:
    quantity = ensure_number(value, "quantity")
    if quantity <= 0:
        raise ValueError(f"'quantity' must be greater than 0, got {value!r}")
    return quantity


def normalize_order(
    raw_order: Dict[str, Any],
    index: int,
    external_id_prefix: str,
    default_currency: str,
    method_codes: set[str],
    default_method: Optional[str],
    type_codes: set[str],
    default_type: Optional[str],
    status_codes: set[str],
    default_status: Optional[str],
) -> Tuple[Dict[str, Any], List[str]]:
    warnings: List[str] = []
    order: Dict[str, Any] = {}

    order["externalId"] = raw_order.get("externalId") or f"{external_id_prefix}-{index:04d}"
    order["currency"] = default_currency
    raw_currency = str(raw_order.get("currency") or "").strip().upper()
    if raw_currency and raw_currency != default_currency:
        warnings.append(
            f"Currency '{raw_currency}' replaced with '{default_currency}'."
        )

    for field in ("firstName", "lastName", "phone", "email", "customerComment"):
        if raw_order.get(field):
            order[field] = raw_order[field]

    raw_method = raw_order.get("orderMethod")
    if raw_method in method_codes:
        order["orderMethod"] = raw_method
    elif default_method:
        order["orderMethod"] = default_method
        if raw_method:
            warnings.append(
                f"Unknown orderMethod '{raw_method}', fallback to '{default_method}'."
            )

    raw_type = raw_order.get("orderType")
    if raw_type in type_codes:
        order["orderType"] = raw_type
    elif default_type:
        order["orderType"] = default_type
        if raw_type:
            warnings.append(
                f"Unknown orderType '{raw_type}', fallback to '{default_type}'."
            )

    raw_status = raw_order.get("status")
    if raw_status in status_codes:
        order["status"] = raw_status
    elif default_status:
        order["status"] = default_status
        if raw_status:
            warnings.append(
                f"Unknown status '{raw_status}', fallback to '{default_status}'."
            )

    items = raw_order.get("items")
    if not isinstance(items, list) or not items:
        raise ValueError("Order must contain a non-empty 'items' array.")

    normalized_items: List[Dict[str, Any]] = []
    for item_index, raw_item in enumerate(items, start=1):
        if not isinstance(raw_item, dict):
            raise ValueError(f"Item #{item_index} must be an object.")

        product_name = raw_item.get("productName")
        if not product_name:
            raise ValueError(f"Item #{item_index} must contain 'productName'.")

        normalized_items.append(
            {
                "productName": product_name,
                "quantity": ensure_positive_quantity(raw_item.get("quantity")),
                "initialPrice": ensure_number(
                    raw_item.get("initialPrice"), "initialPrice"
                ),
            }
        )

    order["items"] = normalized_items

    if isinstance(raw_order.get("delivery"), dict):
        order["delivery"] = raw_order["delivery"]

    if isinstance(raw_order.get("customFields"), dict):
        order["customFields"] = raw_order["customFields"]

    return order, warnings


def iter_orders(data: Any) -> Iterable[Dict[str, Any]]:
    if not isinstance(data, list):
        raise ValueError("Orders file must contain a JSON array.")

    for item in data:
        if not isinstance(item, dict):
            raise ValueError("Each order must be a JSON object.")
        yield item


def main() -> int:
    args = parse_args()
    env = load_env(Path(args.env_file))

    try:
        base_url = env["RETAILCRM_URL"]
        api_key = env["RETAILCRM_API_KEY"]
    except KeyError as exc:
        print(f"Missing required env variable: {exc.args[0]}", file=sys.stderr)
        return 1

    orders_file = Path(args.file or env.get("ORDERS_FILE", "mock_orders.json"))
    if not orders_file.exists():
        print(f"Orders file not found: {orders_file}", file=sys.stderr)
        return 1

    external_id_prefix = env.get("ORDER_EXTERNAL_ID_PREFIX", "mock-order")
    default_currency = (env.get("ORDER_CURRENCY") or "KZT").strip().upper()
    timeout_seconds = int(env.get("REQUEST_TIMEOUT_SECONDS", "30"))
    delay_seconds = (
        args.delay
        if args.delay is not None
        else float(env.get("REQUEST_DELAY_SECONDS", "0.15"))
    )

    if not default_currency:
        print("ORDER_CURRENCY must not be empty.", file=sys.stderr)
        return 1

    client = RetailCrmClient(base_url=base_url, api_key=api_key, timeout=timeout_seconds)

    try:
        sites = client.get("/reference/sites")["sites"]
        order_methods = client.get("/reference/order-methods")["orderMethods"]
        order_types = client.get("/reference/order-types")["orderTypes"]
        statuses = client.get("/reference/statuses")["statuses"]
        site_code = pick_site_code(env.get("RETAILCRM_SITE_CODE"), sites)
    except Exception as exc:
        print(f"Failed to load RetailCRM reference data: {exc}", file=sys.stderr)
        return 1

    method_codes = active_codes(order_methods)
    type_codes = active_codes(order_types)
    status_codes = active_codes(statuses)

    default_method = pick_default_code(order_methods)
    default_type = pick_default_code(order_types)
    default_status = "new" if "new" in status_codes else pick_default_code(statuses)

    try:
        raw_orders = list(iter_orders(load_json(orders_file)))
    except Exception as exc:
        print(f"Failed to read orders file: {exc}", file=sys.stderr)
        return 1

    if args.limit is not None:
        raw_orders = raw_orders[: args.limit]

    uploaded = 0
    failed = 0

    for index, raw_order in enumerate(raw_orders, start=1):
        try:
            order, warnings = normalize_order(
                raw_order=raw_order,
                index=index,
                external_id_prefix=external_id_prefix,
                default_currency=default_currency,
                method_codes=method_codes,
                default_method=default_method,
                type_codes=type_codes,
                default_type=default_type,
                status_codes=status_codes,
                default_status=default_status,
            )
        except Exception as exc:
            failed += 1
            print(f"[{index}] Validation failed: {exc}", file=sys.stderr)
            continue

        for warning in warnings:
            print(f"[{index}] Warning: {warning}", file=sys.stderr)

        if args.dry_run:
            print(json.dumps({"site": site_code, "order": order}, ensure_ascii=False))
            uploaded += 1
            continue

        try:
            payload = {
                "site": site_code,
                "order": json.dumps(order, ensure_ascii=False),
            }
            try:
                response = client.post_form("/orders/create", payload)
                if not response.get("success"):
                    raise RuntimeError(response)

                uploaded += 1
                print(
                    f"[{index}] Uploaded externalId={order['externalId']} "
                    f"id={response.get('id')}"
                )
            except RetailCrmApiError as exc:
                if exc.details.get("errorMsg") != "Order already exists.":
                    raise

                response = client.post_form(
                    f"/orders/{quote(str(order['externalId']), safe='')}/edit",
                    payload,
                )
                if not response.get("success"):
                    raise RuntimeError(response)

                uploaded += 1
                print(
                    f"[{index}] Updated externalId={order['externalId']} "
                    f"id={response.get('id')}"
                )
        except Exception as exc:
            failed += 1
            print(
                f"[{index}] Upload failed for externalId={order['externalId']}: {exc}",
                file=sys.stderr,
            )

        if delay_seconds > 0:
            time.sleep(delay_seconds)

    print(
        f"Done. Processed={len(raw_orders)} uploaded={uploaded} failed={failed}",
        file=sys.stderr if failed else sys.stdout,
    )

    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: upload_orders_to_retailcrm.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

export interface Logger {
  log(message: string): void;
  error(message: string): void;
}

export interface UploadOptions {
  envFile?: string;
  file?: string;
  dryRun?: boolean;
  limit?: number;
  delay?: number;
}

// ----- Utilities -----

function loadEnv(envFile = ".env"): Record<string, string> {
  const result: Record<string, string> = {};
  if (fs.existsSync(envFile)) {
    const lines = fs.readFileSync(envFile, "utf-8").split("\n");
    for (const rawLine of lines) {
      const line = rawLine.trim();
      if (!line || line.startsWith("#") || !line.includes("=")) continue;
      const eqIdx = line.indexOf("=");
      const key = line.slice(0, eqIdx).trim();
      let value = line.slice(eqIdx + 1).trim();
      if (
        (value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))
      ) {
        value = value.slice(1, -1);
      }
      result[key] = value;
    }
  }
  for (const [key, value] of Object.entries(process.env)) {
    if (value) result[key] = value;
  }
  return result;
}

function loadJson(filePath: string): unknown {
  // turbopackIgnore: true — path is dynamic, resolved from env/options at runtime
  const buffer = fs.readFileSync(/* turbopackIgnore: true */ filePath);
  // Strip UTF-8 BOM if present
  const text =
    buffer[0] === 0xef && buffer[1] === 0xbb && buffer[2] === 0xbf
      ? buffer.slice(3).toString("utf-8")
      : buffer.toString("utf-8");
  return JSON.parse(text);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ----- HTTP helper -----

async function fetchWithTimeout(
  url: string,
  options: RequestInit,
  timeoutMs: number,
): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...options, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

// ----- RetailCRM client -----

class RetailCrmApiError extends Error {
  constructor(
    message: string,
    public details: Record<string, unknown>,
  ) {
    super(message);
    this.name = "RetailCrmApiError";
  }
}

class RetailCrmClient {
  private baseUrl: string;

  constructor(
    baseUrl: string,
    private apiKey: string,
    private timeout: number = 30,
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async get(apiPath: string): Promise<Record<string, unknown>> {
    const response = await fetchWithTimeout(
      `${this.baseUrl}${apiPath}`,
      { headers: { "X-API-KEY": this.apiKey, Accept: "application/json" } },
      this.timeout * 1000,
    );
    return this._parseResponse(response);
  }

  async postForm(
    apiPath: string,
    payload: Record<string, string>,
  ): Promise<Record<string, unknown>> {
    const response = await fetchWithTimeout(
      `${this.baseUrl}${apiPath}`,
      {
        method: "POST",
        headers: {
          "X-API-KEY": this.apiKey,
          Accept: "application/json",
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams(payload).toString(),
      },
      this.timeout * 1000,
    );
    return this._parseResponse(response);
  }

  private async _parseResponse(
    response: Response,
  ): Promise<Record<string, unknown>> {
    const text = await response.text();
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(text) as Record<string, unknown>;
    } catch {
      throw new RetailCrmApiError(
        `RetailCRM API returned non-JSON: ${text.slice(0, 500)}`,
        { errorMsg: text.slice(0, 500) },
      );
    }

    if (!response.ok) {
      throw new RetailCrmApiError(
        `RetailCRM API returned HTTP ${response.status}: ${JSON.stringify(data)}`,
        data,
      );
    }

    return data;
  }
}

// ----- Reference data helpers -----

function pickSiteCode(
  explicit: string | undefined,
  sites: Record<string, Record<string, unknown>>,
): string {
  if (explicit) {
    if (explicit in sites) return explicit;
    throw new Error(
      `Configured RETAILCRM_SITE_CODE '${explicit}' is not available.`,
    );
  }
  const keys = Object.keys(sites);
  if (keys.length === 1) return keys[0];
  throw new Error(
    "RetailCRM account has multiple sites. Set RETAILCRM_SITE_CODE in .env.",
  );
}

function activeCodes(
  reference: Record<string, Record<string, unknown>>,
): Set<string> {
  const result = new Set<string>();
  for (const [code, item] of Object.entries(reference)) {
    if (item.active !== false) result.add(code);
  }
  return result;
}

function pickDefaultCode(
  reference: Record<string, Record<string, unknown>>,
): string | null {
  for (const [code, item] of Object.entries(reference)) {
    if (item.defaultForApi) return code;
  }
  for (const [code, item] of Object.entries(reference)) {
    if (item.defaultForCrm) return code;
  }
  for (const [code, item] of Object.entries(reference)) {
    if (item.active !== false) return code;
  }
  return null;
}

// ----- Order normalization -----

function ensureNumber(value: unknown, fieldName: string): number {
  const n = Number(value);
  if (Number.isNaN(n)) {
    throw new Error(
      `'${fieldName}' must be a number, got ${JSON.stringify(value)}`,
    );
  }
  return n;
}

function ensurePositiveQuantity(value: unknown): number {
  const qty = ensureNumber(value, "quantity");
  if (qty <= 0) {
    throw new Error(
      `'quantity' must be greater than 0, got ${JSON.stringify(value)}`,
    );
  }
  return qty;
}

interface NormalizedItem {
  productName: string;
  quantity: number;
  initialPrice: number;
}

interface NormalizedOrder {
  externalId: string;
  currency: string;
  orderMethod?: string;
  orderType?: string;
  status?: string;
  items: NormalizedItem[];
  firstName?: string;
  lastName?: string;
  phone?: string;
  email?: string;
  customerComment?: string;
  delivery?: Record<string, unknown>;
  customFields?: Record<string, unknown>;
}

function normalizeOrder(
  rawOrder: Record<string, unknown>,
  index: number,
  externalIdPrefix: string,
  defaultCurrency: string,
  methodCodes: Set<string>,
  defaultMethod: string | null,
  typeCodes: Set<string>,
  defaultType: string | null,
  statusCodes: Set<string>,
  defaultStatus: string | null,
): [NormalizedOrder, string[]] {
  const warnings: string[] = [];
  const order: Partial<NormalizedOrder> & Record<string, unknown> = {};

  order.externalId =
    (rawOrder.externalId as string) ||
    `${externalIdPrefix}-${String(index).padStart(4, "0")}`;

  order.currency = defaultCurrency;
  const rawCurrency = String(rawOrder.currency ?? "")
    .trim()
    .toUpperCase();
  if (rawCurrency && rawCurrency !== defaultCurrency) {
    warnings.push(
      `Currency '${rawCurrency}' replaced with '${defaultCurrency}'.`,
    );
  }

  for (const field of [
    "firstName",
    "lastName",
    "phone",
    "email",
    "customerComment",
  ]) {
    if (rawOrder[field]) order[field] = rawOrder[field];
  }

  const rawMethod = rawOrder.orderMethod as string | undefined;
  if (rawMethod && methodCodes.has(rawMethod)) {
    order.orderMethod = rawMethod;
  } else if (defaultMethod) {
    order.orderMethod = defaultMethod;
    if (rawMethod) {
      warnings.push(
        `Unknown orderMethod '${rawMethod}', fallback to '${defaultMethod}'.`,
      );
    }
  }

  const rawType = rawOrder.orderType as string | undefined;
  if (rawType && typeCodes.has(rawType)) {
    order.orderType = rawType;
  } else if (defaultType) {
    order.orderType = defaultType;
    if (rawType) {
      warnings.push(
        `Unknown orderType '${rawType}', fallback to '${defaultType}'.`,
      );
    }
  }

  const rawStatus = rawOrder.status as string | undefined;
  if (rawStatus && statusCodes.has(rawStatus)) {
    order.status = rawStatus;
  } else if (defaultStatus) {
    order.status = defaultStatus;
    if (rawStatus) {
      warnings.push(
        `Unknown status '${rawStatus}', fallback to '${defaultStatus}'.`,
      );
    }
  }

  const items = rawOrder.items;
  if (!Array.isArray(items) || items.length === 0) {
    throw new Error("Order must contain a non-empty 'items' array.");
  }

  const normalizedItems: NormalizedItem[] = [];
  for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
    const rawItem = items[itemIndex];
    if (typeof rawItem !== "object" || rawItem === null) {
      throw new Error(`Item #${itemIndex + 1} must be an object.`);
    }
    const ri = rawItem as Record<string, unknown>;
    const productName = ri.productName;
    if (!productName) {
      throw new Error(`Item #${itemIndex + 1} must contain 'productName'.`);
    }
    normalizedItems.push({
      productName: String(productName),
      quantity: ensurePositiveQuantity(ri.quantity),
      initialPrice: ensureNumber(ri.initialPrice, "initialPrice"),
    });
  }

  order.items = normalizedItems;

  if (
    typeof rawOrder.delivery === "object" &&
    rawOrder.delivery !== null &&
    !Array.isArray(rawOrder.delivery)
  ) {
    order.delivery = rawOrder.delivery as Record<string, unknown>;
  }

  if (
    typeof rawOrder.customFields === "object" &&
    rawOrder.customFields !== null &&
    !Array.isArray(rawOrder.customFields)
  ) {
    order.customFields = rawOrder.customFields as Record<string, unknown>;
  }

  return [order as NormalizedOrder, warnings];
}

// ----- Main exported run function -----

export async function run(
  options: Partial<UploadOptions> = {},
  logger: Logger = { log: console.log, error: console.error },
): Promise<{ ok: boolean }> {
  const envFile = options.envFile ?? ".env";
  const env = loadEnv(envFile);

  const baseUrl = env.RETAILCRM_URL;
  const apiKey = env.RETAILCRM_API_KEY;

  if (!baseUrl) {
    logger.error("Missing required env variable: RETAILCRM_URL");
    return { ok: false };
  }
  if (!apiKey) {
    logger.error("Missing required env variable: RETAILCRM_API_KEY");
    return { ok: false };
  }

  const ordersFilePath = options.file ?? env.ORDERS_FILE ?? "mock_orders.json";
  if (!fs.existsSync(ordersFilePath)) {
    logger.error(`Orders file not found: ${ordersFilePath}`);
    return { ok: false };
  }

  const externalIdPrefix = env.ORDER_EXTERNAL_ID_PREFIX ?? "mock-order";
  const defaultCurrency = (env.ORDER_CURRENCY ?? "KZT").trim().toUpperCase();
  const timeoutSeconds = parseInt(env.REQUEST_TIMEOUT_SECONDS ?? "30", 10);
  const delaySeconds =
    options.delay !== undefined
      ? options.delay
      : parseFloat(env.REQUEST_DELAY_SECONDS ?? "0.15");
  const dryRun = options.dryRun ?? false;
  const limit = options.limit ?? null;

  if (!defaultCurrency) {
    logger.error("ORDER_CURRENCY must not be empty.");
    return { ok: false };
  }

  const client = new RetailCrmClient(baseUrl, apiKey, timeoutSeconds);

  let sites: Record<string, Record<string, unknown>>;
  let orderMethods: Record<string, Record<string, unknown>>;
  let orderTypes: Record<string, Record<string, unknown>>;
  let statuses: Record<string, Record<string, unknown>>;
  let siteCode: string;

  try {
    const sitesResp = await client.get("/reference/sites");
    sites = sitesResp.sites as Record<string, Record<string, unknown>>;

    const methodsResp = await client.get("/reference/order-methods");
    orderMethods = methodsResp.orderMethods as Record<
      string,
      Record<string, unknown>
    >;

    const typesResp = await client.get("/reference/order-types");
    orderTypes = typesResp.orderTypes as Record<
      string,
      Record<string, unknown>
    >;

    const statusesResp = await client.get("/reference/statuses");
    statuses = statusesResp.statuses as Record<string, Record<string, unknown>>;

    siteCode = pickSiteCode(env.RETAILCRM_SITE_CODE, sites);
  } catch (err) {
    logger.error(
      `Failed to load RetailCRM reference data: ${err instanceof Error ? err.message : String(err)}`,
    );
    return { ok: false };
  }

  const methodCodes = activeCodes(orderMethods);
  const typeCodes = activeCodes(orderTypes);
  const statusCodes = activeCodes(statuses);

  const defaultMethod = pickDefaultCode(orderMethods);
  const defaultType = pickDefaultCode(orderTypes);
  const defaultStatus = statusCodes.has("new")
    ? "new"
    : pickDefaultCode(statuses);

  let rawOrders: Record<string, unknown>[];
  try {
    const data = loadJson(ordersFilePath);
    if (!Array.isArray(data)) {
      throw new Error("Orders file must contain a JSON array.");
    }
    for (const item of data) {
      if (typeof item !== "object" || item === null) {
        throw new Error("Each order must be a JSON object.");
      }
    }
    rawOrders = data as Record<string, unknown>[];
  } catch (err) {
    logger.error(
      `Failed to read orders file: ${err instanceof Error ? err.message : String(err)}`,
    );
    return { ok: false };
  }

  if (limit !== null) {
    rawOrders = rawOrders.slice(0, limit);
  }

  let uploaded = 0;
  let failed = 0;

  for (let i = 0; i < rawOrders.length; i++) {
    const index = i + 1;
    const rawOrder = rawOrders[i];

    let order: NormalizedOrder;
    let warnings: string[];

    try {
      [order, warnings] = normalizeOrder(
        rawOrder,
        index,
        externalIdPrefix,
        defaultCurrency,
        methodCodes,
        defaultMethod,
        typeCodes,
        defaultType,
        statusCodes,
        defaultStatus,
      );
    } catch (err) {
      failed++;
      logger.error(
        `[${index}] Validation failed: ${err instanceof Error ? err.message : String(err)}`,
      );
      continue;
    }

    for (const warning of warnings) {
      logger.error(`[${index}] Warning: ${warning}`);
    }

    if (dryRun) {
      logger.log(JSON.stringify({ site: siteCode, order }));
      uploaded++;
      continue;
    }

    try {
      const payload = {
        site: siteCode,
        order: JSON.stringify(order),
      };

      try {
        const response = await client.postForm("/orders/create", payload);
        if (!response.success) throw new Error(JSON.stringify(response));

        uploaded++;
        logger.log(
          `[${index}] Uploaded externalId=${order.externalId} id=${response.id}`,
        );
      } catch (err) {
        if (
          err instanceof RetailCrmApiError &&
          err.details.errorMsg === "Order already exists."
        ) {
          const editPath = `/orders/${encodeURIComponent(String(order.externalId))}/edit`;
          const response = await client.postForm(editPath, payload);
          if (!response.success) throw new Error(JSON.stringify(response));

          uploaded++;
          logger.log(
            `[${index}] Updated externalId=${order.externalId} id=${response.id}`,
          );
        } else {
          throw err;
        }
      }
    } catch (err) {
      failed++;
      logger.error(
        `[${index}] Upload failed for externalId=${order.externalId}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    if (delaySeconds > 0) await sleep(delaySeconds * 1000);
  }

  const summary = `Done. Processed=${rawOrders.length} uploaded=${uploaded} failed=${failed}`;
  if (failed > 0) {
    logger.error(summary);
  } else {
    logger.log(summary);
  }

  return { ok: failed === 0 };
}

// ----- CLI entry point -----

function parseArgs(): Partial<UploadOptions> {
  const args = process.argv.slice(2);
  const opts: Partial<UploadOptions> = {};
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "--env-file":
        opts.envFile = args[++i];
        break;
      case "--file":
        opts.file = args[++i];
        break;
      case "--dry-run":
        opts.dryRun = true;
        break;
      case "--limit":
        opts.limit = parseInt(args[++i], 10);
        break;
      case "--delay":
        opts.delay = parseFloat(args[++i]);
        break;
    }
  }
  return opts;
}

async function main(): Promise<void> {
  const { ok } = await run(parseArgs(), {
    log: console.log,
    error: console.error,
  });
  process.exit(ok ? 0 : 1);
}

const __filename = fileURLToPath(import.meta.url);
if (process.argv[1] && path.resolve(process.argv[1]) === __filename) {
  main().catch((err: unknown) => {
    console.error(err);
    process.exit(1);
  });
}
